src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -430,6 +430,22 @@ def get_installation_error(self, e):
                 "Could not find a version that satisfies the requirement %s",
                 req_disp,
             )
+
+            acands = self._finder.find_all_candidates(req.project_name)
+            aversions = set(str(cand.version) for cand in acands)
+            afrom = set(str(cand.link.comes_from) for cand in acands)
+            cache_dir = (self._wheel_cache.cache_dir
+                         if self._wheel_cache is not None else None)
+            diagnostic_hint = (
+                f"\nAvailable versions for '{req.project_name}':\n"
+                f"{', '.join(aversions) if aversions else '(none)'}\n\n"
+                f"Found in index{'es' if len(afrom) > 1 else ''}:\n"
+                f"{', '.join(afrom) if afrom else '(none)'}\n\n"
+                f"Cache dir:\n"
+                f"{cache_dir or '(not used)'}"
+            )
+            logger.warning(diagnostic_hint)
```

Comment:
The approach lgtm, but the message formatting is much too complicated imo. I would simplify this to something like

```
Available versions: <list of versions joined by comma>
Index used:
  <list of index URLs joined by newline and indented>
```

Some points to note:

* The project name is already shown in the message directly above and would be duplicated info here.
* Available versions need to be sorted to be readable (they are not right now).
* Showing the cache directory is not useful imo. Cache is only a factor during package build and install, not here.

Reply:
Thanks.

> The project name is already shown in the message directly above and would be duplicated info here.
Correct. Removed

> Available versions need to be sorted to be readable (they are not right now).
Added sorting to versions using `distutils.version.LooseVersion`. Indexes are now sorted too, as to ensure that they don't jump around on different runs.

> Showing the cache directory is not useful imo.
Yea, I guess even if it was, you might as well just do `pip cache info`. Having it in every one of these outputs is just too verbose anyway. I removed it.

> `comes_from` is not necessarily an index page.
Huh? What else could it be at that point? Replaced `Index(es):` with `Sources:`.

This is how it looks right now:
![after](https://user-images.githubusercontent.com/6691643/103446137-9be40600-4c7c-11eb-87a3-fc78e20fa749.png)

Reply:
I've also removed the empty newlines, so now it looks like this:
![after2](https://user-images.githubusercontent.com/6691643/103446482-7ce77300-4c80-11eb-87d3-e8c1bee6f748.png)

Reply:
Instead of making this a separate logging.warning call, I think we should be adding this information to the original logging.critical call.

Reply:
> Added sorting to versions using `distutils.version.LooseVersion`

We should use the version type from `packaging`. I don't think we should add a dependency on distutils here, as Python core is planning on removing that at some point.

Reply:
@pfmoore I didn't know about that. Will do.
@pradyunsg The point of doing that was to make it different from the error messages, as - technically - these aren't, and for readability. I can change that too, no problem there.

Reply:
> Huh? What else could it be at that point? Replaced `Index(es):` with `Sources:`.

PackageFinder also looks at find-links pages, which are technically not package indexes.

This reminds me though, we should add a check and display this whole message only when the user did not specify any direct URL requirements on the package. It would be quite confusing IMO if the user specify `foo @ git+https://github.com/...` and pip displays this.

Reply:
@pfmoore done
@pradyunsg Done. This is a comparison of the before / after:
![2021-01-02_15-34](https://user-images.githubusercontent.com/6691643/103459468-fed2ad00-4d0f-11eb-973e-ce947110693e.png)

Reply:
@uranusjr The sources section has been removed. Not sure how to go about the `foo@git+...` thing though.